File: 프로그래머스/lv2/1844. 게임 맵 최단거리/게임 맵 최단거리.js
function solution(maps) {
  const n = maps.length;
  const m = maps[0].length;
  const dist = Array(n)
    .fill()
    .map(() => Array(m).fill(-1));
  const queue = [];
  const dx = [0, 1, 0, -1];
  const dy = [1, 0, -1, 0];
  dist[0][0] = 1;
  queue.push([0, 0]);

  while (queue.length !== 0) {
    const [x, y] = queue.shift();
    if ([x, y] === [n - 1, m - 1]) return dist[x][y];
    for (let dir = 0; dir < 4; dir++) {
      const newX = x + dx[dir];
      const newY = y + dy[dir];
      if (newX < 0 || newX >= n) continue;
      if (newY < 0 || newY >= m) continue;
      if (dist[newX][newY] !== -1) continue;
      if (maps[newX][newY] === 0) continue;
      dist[newX][newY] = dist[x][y] + 1;
      queue.push([newX, newY]);
    }
  }

  return dist[n - 1][m - 1];
}

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque

def solution(maps):   
    def bfs():
        n = len(maps)
        m = len(maps[0])
        visited = [[False] * m for _ in range(n)]
        dirs = [[-1, 0], [0, -1], [1, 0], [0, 1]]
        queue = deque([(0, 0, 1)])
        
        while queue:
            x, y, dist = queue.popleft()
            
            if x == n-1 and y == m-1:
                return dist
            
            for dx, dy in dirs:
                nx = x + dx
                ny = y + dy
                
                if nx >= n or nx < 0 or ny >= m or ny < 0:
                    continue
                if maps[nx][ny] == 0:
                    continue
                if visited[nx][ny]:
                    continue
                
                visited[nx][ny] = True
                queue.append((nx, ny, dist+1))
        
        return -1
        
    return bfs()